### PR TITLE
Raise `FileNotFoundError` in `read_json` if input looks like file path but file is missing

### LINKED ISSUE #29102: Misleading error messages when opening inexistent json file
When opening a json file that doesn't exist with `read_json`, we get a `ValueError` with error messages like:

```python
>>> import pandas
>>> pandas.read_json('no_file.json')
ValueError: Unexpected character found when decoding 'null'

>>> pandas.read_json('a_dir/no_file.json')
ValueError: Expected object or value
```

It'd probably be better to raise an exception like:
```
FileNotFoundError: No such file: 'no_file.json'
```

### PR DESCRIPTION
**Pull request by janosh**
- [x] closes #29102
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/json/_json.py
Comment (twoertwein) [inline pandas/io/json/_json.py:724]: We have a list of all supported compression extensions:
```py
from pandas.io.common import _compression_to_extension

.endswith([".json"] +  [f".json.{compression}" for compression in _compression_to_extension])
```

Comment (janosh) [reply pandas/io/json/_json.py:724]: Oh good, I was looking for that.

File: doc/source/whatsnew/v1.5.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:569]: Technically this is an API change, so could you put this under the `Other API changes` section?

File: pandas/io/json/_json.py
Comment (jreback) [inline pandas/io/json/_json.py:723]: so IIRC we did do this quite awhile ago and in fact just did essentially `file_exists(filepath_or_buffer)` 

I believe this was reverted as it is not reliable and can potentially overflow the file system.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/json/_json.py
Comment (janosh) [reply pandas/io/json/_json.py:723]: @jreback Sounds very odd. How can file existence check go so wrong? Do you remember under what conditions? Was it a rare edge case? Couldn't find any old issue about this. Do you remember a keyword to search for?

Comment (jreback) [reply pandas/io/json/_json.py:723]: if its a string this can simply overflow the file system

Comment (janosh) [reply pandas/io/json/_json.py:723]: So would a check for string length before file_exists help?

```py
isinstance(filepath_or_buffer, str) and len(filepath_or_buffer) < 256 and ... # or whatever is safe
```

Comment (twoertwein) [reply pandas/io/json/_json.py:723]: `file_exists` tries to catch certain errors: 
https://github.com/pandas-dev/pandas/blob/cec066d221cc48a858fcae650d254b5392dd1247/pandas/io/common.py#L1095

Comment (janosh) [reply pandas/io/json/_json.py:723]: @twoertwein @jreback Maybe worth adding a test case for `read_json` that tries to load a very long file path?

Comment (janosh) [reply pandas/io/json/_json.py:723]: Seems like the concern @jreback raised is handled by file_exists. Any other reason not to merge this?

**Review by mroeschke — APPROVED** [review]
LGTM. Will need approval from @jreback to merge. (long string's overflowing the filesystem is handled by `file_exists`; there's a comment `# gh-5874: if the filepath is too long will raise here`)

**Review by mroeschke — CHANGES_REQUESTED** [review]
Actually, after merging master looks like there's a code path needs updating.

Also @twoertwein, for other file extensions with compression, it doesn't appear we manually raise a `FileNotFoundError`. Should this automatically be handled by the handler code?

**Comment by twoertwein** [comment]
> Also @twoertwein, for other file extensions with compression, it doesn't appear we manually raise a `FileNotFoundError`. Should this automatically be handled by the handler code?


Would maybe be better to raise here when the file doesn't exist:

https://github.com/pandas-dev/pandas/blob/b1b788fefcdc1f46cebbf6a0df4a9f321ca20086/pandas/io/common.py#L692

File: doc/source/whatsnew/v1.5.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:435]: move to the I/O section

File: pandas/io/json/_json.py
Comment (jreback) [reply pandas/io/json/_json.py:723]: ok this looks fine, can we do exactly that add a superlong json string here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.5.0.rst
Comment (mroeschke) [reply doc/source/whatsnew/v1.5.0.rst:435]: I think this is technically an API change since the exception class changed.

Comment (janosh) [reply doc/source/whatsnew/v1.5.0.rst:435]: It was under whatsnew **IO** originally but moved to **API changes** following https://github.com/pandas-dev/pandas/pull/46718#discussion_r846662440.

**Comment by janosh** [comment]
[Is this a known error](https://github.com/pandas-dev/pandas/runs/6754333579?check_suite_focus=true#step:9:62) in `TestiLocBaseIndependent.test_iloc_setitem_multicolumn_to_datetime`? Looks unrelated to my changes.

**Review by jreback — APPROVED** [review]
lgtm merge on green

**Comment by jreback** [comment]
cc @mroeschke

File: pandas/io/json/_json.py
Comment (twoertwein) [reply pandas/io/json/_json.py:730]: After read_json knows that it is a file, I think you can just raise inside get_handle when the file doesn't exist:

```py
if "r" not in mode and is_path and file_exists(handle):
    raise ...
```

Comment (janosh) [reply pandas/io/json/_json.py:730]: You mean this, right?
```py
if "r" in mode and is_path and not file_exists(handle):
    raise ...
```

Comment (janosh) [reply pandas/io/json/_json.py:730]: Actually `read_json` doesn't even call `get_handle` since it's behind a `file_exists` check:

```py
if (
    not isinstance(filepath_or_buffer, str)
    or is_url(filepath_or_buffer)
    or is_fsspec_url(filepath_or_buffer)
    or file_exists(filepath_or_buffer)
):
    self.handles = get_handle(...)
```

Comment (twoertwein) [reply pandas/io/json/_json.py:730]: I think there were attempts at creating an efficient way of checking whether a str is a legit JSON string. I assume these attempts were unsuccessful - we check the reverse. If that's the case, then moving the check into get_handle, will not work.

Comment (janosh) [reply pandas/io/json/_json.py:730]: > If that's the case, then moving the check into get_handle, will not work.

So leave PR as is?

Comment (twoertwein) [reply pandas/io/json/_json.py:730]: > So leave PR as is?

Probably.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Great, thanks @janosh

**Pull request merged**
